rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,713 @@
+it could be technically possible but appear as a very ugly attribute usage, which could make the
+source muddy and unclear.
+
+Additionally, while derive macros are written in Rust code, which can be much easier to grok than
+declarative macros, it represents a section of code that may be harder to understand and modify than
+if we used a more brute force approach.
+
+## Prior Art
+
+Notably, and perhaps most obviously, Cue itself can be used for data validation purposes. There are
+also other projects that slot themselves into the same general domain, such as [OPA/Rego][opa_rego]
+and [CDDL][cddl].
+
+The primary problem with all of these solutions is that they're all custom languages, with far less
+cross-platform support, and a far steeper learning curve. Even though our approach generally
+concerns itself with obfuscating the schema tool itself, and focusing on making it trivially to
+generate the schema from annotated Rust types/fields, eventually the rubber must meet the road, and
+this is where these other tools would fall down for our case and become very hard to wield
+correctly.
+
+## Alternatives
+
+While `schemars` itself has support for annotating Rust types in almost the exact same way as we've
+proposed above, it lacks a few features necessary for our use case:
+
+- no support for the serde `alias` attribute feature
+- no support for defining generic metadata for a type/field and exposing it in the schema
+- no mechanism to override constraints for a field which already defines its own constraints at the
+  type level
+
+The lack of these features make it much harder to correctly generate our schema, as we would still
+be required to do the minimum amount of work to support defining custom attributes for the missing
+features, and the work to support at least one attribute is 90% of the work to support two, or
+three, or more, custom attributes. It would also mean that developers would need to figure out
+whether or not `schemars` provided a certain behavior via its attributes, versus being able to only
+need to remember how to get to the documentation for the our proposed internal approach.
+
+## Outstanding Questions
+
+- Can the proposed solution work seamlessly for all types that we use in our configuration? (For
+  example, `Option<T>` or any other type that needs to partially forward its declaration to the
+  underlying `T`)
+- Can downstream consumers extract the necessary information purely from the specified JSONSchema
+  fields and any additional custom metadata as string key/value pairs?  Do they need a richer
+  representation of custom metadata?
+- Is there a better way, or a way at all, to handle overriding subfields at compile-time?
+  (Procedural macros operate on the Rust AST, so for example, we cannot annotate a field that points
+  to type B and reference a field that only exists on B, so we have to generate code such that it
+  eventually fails at run-time.)
```

Comment:
Could you give an example of this kind of problem? It sounds a bit more complex than something like overriding default batch sizes, but beyond that I'm not super clear.

Reply:
So take the example of batch settings, but let's focus on a particular use case: setting limits on the number of events in a sink batch.  We might want to do something like:

```rust
struct SinkConfig {
  // Allow up to 1000 events per batch... but woops, we used the wrong field name!
  #[configurable(field(max_eventsz, max(1000)))]
  batch: BatchConfig,
}
```

In this, we're referencing a field (`max_eventz`) on a type (`BatchConfig`) that the procedural macro can't "see": we can generate code that references `self.batch.max_eventz` trivially, but we don't know enough about `BatchConfig` -- because we're operating on `SinkConfig` at this point, from the macro's perspective -- to generate an error at the callsite of the attribute to say "the field `max_eventz` does not exist, did you mean <list of real fields on BatchConfig>?".

We can generate code that will fail to compile, but it won't be a contextual callsite error at the macro, just a generic error.  Well, worst case it'll be a generic error.  Maybe there's still a way to provide enough information that the compiler can attribute the error to that general block, but just not as contextual as I wish it could be.

Reply:
Got it, that makes sense.